lib/engine/action/place_token.rb

```diff
@@ -5,22 +5,26 @@
 module Engine
   module Action
     class PlaceToken < Base
-      attr_reader :city, :slot
+      attr_reader :city, :slot, :token
 
-      def initialize(entity, city, slot)
+      def initialize(entity, city, slot, token_id = nil)
         @entity = entity
         @city = city
         @slot = slot
+        # After processing the action the token may no longer be in the entity
+        @token_id = token_id
```

Comment:
no reason to store this, we should ensure that tokens maitain the id, add id to token and init with index,